horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/data/ExportedLabelValues.java

```diff
@@ -36,25 +40,91 @@ public ExportedLabelValues(LabelValueMap v, Integer runId, Integer datasetId, In
-        return fps;
+
+        Map<RunDatasetKey, ExportedLabelValues> exportedLabelValues = nodes.stream()
+                .collect(Collectors.groupingBy(
+                        objects -> { // Create the key as a combination of runId and datasetId
+                            Integer runId = (Integer) objects[2];
+                            Integer datasetId = (Integer) objects[3];
+                            return new RunDatasetKey(runId, datasetId);
+                        },
+                        LinkedHashMap::new, // preserve the order in which I receive the nodes
+                        Collectors.collectingAndThen(
+                                Collectors.toList(), // collect each group (by runId and datasetId) into a list of elements
+                                (groupedObjects) -> { // process each group and create the corresponding ExportedLabelValues
+                                    ExportedLabelValues exportedLabelValue = getExportedLabelValues(groupedObjects);
+
+                                    groupedObjects.forEach(objects -> {
+                                        // skip records where the labelName is null
+                                        if (objects[0] != null) {
+                                            String labelName = (String) objects[0];
+                                            JsonNode node = (JsonNode) objects[1];
+                                            exportedLabelValue.values.put(labelName, node);
+                                        }
+                                    });
+
+                                    return exportedLabelValue;
+                                })));
+
+        return new ArrayList<>(exportedLabelValues.values());
+    }
+
+    private static ExportedLabelValues getExportedLabelValues(List<Object[]> groupedObjects) {
+        Integer runId = (Integer) groupedObjects.get(0)[2];
+        Integer datasetId = (Integer) groupedObjects.get(0)[3];
+        Instant start = (Instant) groupedObjects.get(0)[4];
+        Instant stop = (Instant) groupedObjects.get(0)[5];
+        return new ExportedLabelValues(new LabelValueMap(), runId, datasetId, start, stop);
+    }
+
+    public static class RunDatasetKey {
+        Integer runId;
+        Integer datasetId;
+
+        public RunDatasetKey(Integer runId, Integer datasetId) {
+            this.runId = runId;
+            this.datasetId = datasetId;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(runId, datasetId);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+
+            RunDatasetKey that = (RunDatasetKey) o;
+
+            if (!Objects.equals(runId, that.runId))
+                return false;
+            return Objects.equals(datasetId, that.datasetId);
+        }
```

Comment:
I just realised, this is an implementation detail that is leaking from the API. A dev can add a java dependency on `horreum-client` and then write something like;

```
import io.hyperfoil.tools.HorreumClient;
import io.hyperfoil.tools.horreum.api.data.ExportedLabelValues;

public class TestClient {

    public void testClientAPI() {
        HorreumClient horreumClient = new HorreumClient.Builder()
                .horreumUrl("http://localhost:".concat(System.getProperty("quarkus.http.test-port")))
                .horreumUser("test")
                .horreumPassword("password")
                .build();

        assert(horreumClient != null);

        ExportedLabelValues.parse(null);

    }
}
```

I think this static method would be better placed in the `LabelValuesService` class

Reply:
Hey @johnaohara, good catch!!!

We shouldn't expose such static methods in the api.

Done with https://github.com/Hyperfoil/Horreum/pull/1977/commits/19a080d3c3c4d1f0272ec5f82d1c1f02af99e701